File: 신고결과받기/javascript/20221108/src/solution2.js
function test() {
  console.log(`distinctReports(["ryan con", "ryan con", "ryan con", "ryan con"]): ["ryan con"] = ${distinctReports(['ryan con', 'ryan con', 'ryan con', 'ryan con'])}`);

  console.log(`splitReports(["ryan con","muzi frodo"]): [[ryan, con],[muzi, frodo]] = ${JSON.stringify(splitReports(['ryan con', 'muzi frodo']))}`);

  console.log(`reportedCounts(["muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi"]): {frodo: 2, neo: 2, muzi:1} = ${[...reportedCounts(['muzi frodo', 'apeach frodo', 'frodo neo', 'muzi neo', 'apeach muzi'])]}`);

  console.log(`mailCounts(["muzi", "frodo", "apeach", "neo"],["muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi"],2): [2,1,1,0] = ${mailCounts(['muzi', 'frodo', 'apeach', 'neo'], ['muzi frodo', 'apeach frodo', 'frodo neo', 'muzi neo', 'apeach muzi'], 2)}`);
}

function solution(id_list, reports, k) {
  // test();
  const answer = mailCounts(id_list, reports, k);
  return answer;
}

function distinctReports(reports) {
  return [...new Set(reports)];
}

function splitReports(reports) {
  return reports.map((report) => report.split(' '));
}

function reportedCounts(reports) {
  const reportedCountsMap = new Map();

  splitReports(distinctReports(reports))
    .forEach(([reporter, reportedUser]) => {
      reportedCountsMap.set(reportedUser, (reportedCountsMap.get(reportedUser) ?? 0) + 1);
    });

  return reportedCountsMap;
}

function mailCounts(id_list, reports, k) {
  const mailCountsMap = new Map();

  const reportedCountsMap = reportedCounts(reports);

  splitReports(distinctReports(reports))
    .forEach(([reporter, reportedUser]) => {
      if (reportedCountsMap.get(reportedUser) >= k) {
        mailCountsMap.set(reporter, (mailCountsMap.get(reporter) ?? 0) + 1);
      }
    });

  return id_list.map((id) => mailCountsMap.get(id) ?? 0);
}

File: 신고결과받기/java/20221107/src/Solution2.java
import java.util.*;
import java.util.stream.*;

class Solution2 {
    public void test() {
        System.out.println("distinctReports(new String[]{\"ryan con\", \"ryan con\", \"ryan con\", \"ryan con\"}): [ryan con] = " + distinctReports(new String[]{"ryan con", "ryan con", "ryan con", "ryan con"}));

        System.out.println("reportedCounts(new String[]{\"muzi frodo\",\"apeach frodo\",\"frodo neo\",\"muzi neo\",\"apeach muzi\"}): {frodo= 2, neo= 2, muzi= 1}" + reportedCounts(new String[]{"muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"}));

        System.out.println("receivedMailCounts(new String[] {\"muzi\", \"frodo\", \"apeach\", \"neo\"}, new String[]{\"muzi frodo\",\"apeach frodo\",\"frodo neo\",\"muzi neo\",\"apeach muzi\"},2): [2,1,1,0]" + Arrays.stream(receivedMailCounts(new String[]{"muzi", "frodo", "apeach", "neo"}, new String[]{"muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"}, 2)).boxed().collect(Collectors.toList()));
    }

    public int[] solution(String[] id_list, String[] reports, int k) {
        // test();
        int[] answer = receivedMailCounts(id_list, reports, k);
        return answer;
    }

    public List<String> distinctReports(String[] reports) {
        return Arrays.stream(reports).distinct().collect(Collectors.toList());
    }

    public Map<String, Integer> reportedCounts(String[] reports) {
        Map<String, Integer> reportedCountsMap = new HashMap<>();

        distinctReports(reports).stream()
                .forEach(report -> {
                    String reporter = report.split(" ")[0];
                    String reportedUser = report.split(" ")[1];

                    reportedCountsMap.put(reportedUser, reportedCountsMap.getOrDefault(reportedUser, 0) + 1);
                });

        return reportedCountsMap;
    }

    public int[] receivedMailCounts(String[] id_list, String[] reports, int k) {
        Map<String, Integer> receivedMailCountsMap = new HashMap<>();
        Map<String, Integer> reportedCountsMap = reportedCounts(reports);

        distinctReports(reports).stream()
                .forEach(report -> {
                    String reporter = report.split(" ")[0];
                    String reportedUser = report.split(" ")[1];

                    if (reportedCountsMap.get(reportedUser) >= k) {
                        receivedMailCountsMap.put(reporter, receivedMailCountsMap.getOrDefault(reporter, 0) + 1);
                    }
                });

        return Arrays.stream(id_list).mapToInt(id -> receivedMailCountsMap.getOrDefault(id, 0)).toArray();
    }
}
